myia/opt/llift.py

```diff
@@ -0,0 +1,82 @@
+"""Lambda lifting."""
+
+from types import SimpleNamespace as NS
+
+from ..info import About
+from ..ir import ANFNode, manage
+from ..utils import OrderedSet, WorkSet
+
+
+def _find_fvs(graph):
+    fvs = OrderedSet()
+    work = WorkSet([graph])
+    for g in work:
+        for fv in g.free_variables_total:
+            if isinstance(fv, ANFNode):
+                fvs.add(fv)
+            else:
+                work.queue(fv)
+    return [fv for fv in fvs if fv.graph not in graph.scope]
+
+
+def lambda_lift(root):
+    """Lambda-lift the graphs that can be lifted.
+
+    Graphs with free variables for which we can identify all calls will be
+    modified to take these free variables as extra arguments.
+
+    This is a destructive operation.
+    """
+    mng = manage(root)
+    graphs = WorkSet(mng.graphs)
+    todo = {}
+
+    for g in graphs:
+        if g.parent and not graphs.processed(g.parent):
+            graphs.requeue(g)
+            continue
+
+        todo[g] = NS(graph=g, calls=[], fvs={})
+
+        if not g.free_variables_total:
+            continue
+        gcalls = [
+            (node, idx)
+            for ct in mng.graph_constants[g]
+            for node, idx in mng.uses[ct]
+        ]
+        if all(idx == 0 for node, idx in gcalls):
+
+            def _param(fv):
+                with About(fv.debug, "llift"):
+                    param = g.add_parameter()
+                    param.abstract = fv.abstract
+                    return param
+
+            todo[g] = NS(
+                graph=g,
+                calls=gcalls,
+                fvs={fv: _param(fv) for fv in _find_fvs(g)},
+            )
+
+    with mng.transact() as tr:
+        for entry in todo.values():
+            # Add the new parameters to all call sites
+            for node, _ in entry.calls:
+                new_node = node.graph.apply(*node.inputs, *entry.fvs)
+                new_node.abstract = node.abstract
+                tr.replace(node, new_node)
+
+    with mng.transact() as tr:
+        for entry in todo.values():
+            # Redirect the fvs to the parameter (those in scope)
+            for fv, param in entry.fvs.items():
+                for node, idx in mng.uses[fv]:
+                    if node.graph in entry.graph.scope:
+                        tr.set_edge(node, idx, param)
```

Comment:
I see that you have a test for multi-level nesting, but I just don't see how it can work.  I probably just don't understand it.

Can you try to explain?

Reply:
I'll try to document what we discussed for future generations:

This is the starting graph:
```
 def f1(x, y):
     def g(z):
         return x + z

     def h():
         return g(y)

     return h()
```

The first loop adds parameters for all the call sites like this:
```
 def f1(x, y):
     def g(z, x_):
         return x + z

     def h(y_, x_):
         return g(y, x)

     return h(y, x)
```

Then the second loop then swaps the freevars for the new args:
```
 def f1(x, y):
     def g(z, x_):
         return x_ + z

     def h(y_, x_):
         return g(y_, x_)

     return h(y, x)
```